data/abi/abi.go

```diff
@@ -0,0 +1,1164 @@
+func MakeDynamicArrayType(argumentType Type) Type {
+	return Type{
+		enumIndex:  ArrayDynamic,
+		childTypes: []Type{argumentType},
+	}
+}
+
+// MakeStringType makes `String` ABI type.
+func MakeStringType() Type {
+	return Type{
+		enumIndex: String,
+	}
+}
+
+// MakeTupleType makes tuple ABI type by taking an array of tuple element types as argument.
+func MakeTupleType(argumentTypes []Type) (Type, error) {
+	if len(argumentTypes) >= (1 << 16) {
+		return Type{}, fmt.Errorf("tuple type child type number >= 2^16 error")
+	}
+	return Type{
+		enumIndex:    Tuple,
+		childTypes:   argumentTypes,
+		staticLength: uint16(len(argumentTypes)),
+	}, nil
+}
+
+// Equal method decides the equality of two types: t == t0.
+func (t Type) Equal(t0 Type) bool {
+	if t.enumIndex != t0.enumIndex {
+		return false
+	} else if t.precision != t0.precision || t.size != t0.size {
+		return false
+	} else if t.staticLength != t0.staticLength {
+		return false
+	} else {
+		if len(t.childTypes) != len(t0.childTypes) {
+			return false
+		}
+		for i := 0; i < len(t.childTypes); i++ {
+			if !t.childTypes[i].Equal(t0.childTypes[i]) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// IsDynamic method decides if an ABI type is dynamic or static.
+func (t Type) IsDynamic() bool {
+	switch t.enumIndex {
+	case ArrayDynamic, String:
+		return true
+	default:
+		for _, childT := range t.childTypes {
+			if childT.IsDynamic() {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+// ByteLen method calculates the byte length of a static ABI type.
+func (t Type) ByteLen() (int, error) {
+	switch t.enumIndex {
+	case Address:
+		return 32, nil
+	case Byte:
+		return 1, nil
+	case Uint, Ufixed:
+		return int(t.size / 8), nil
+	case Bool:
+		return 1, nil
+	case ArrayStatic:
+		if t.childTypes[0].enumIndex == Bool {
+			byteLen := int(t.staticLength) / 8
+			if t.staticLength%8 != 0 {
+				byteLen++
+			}
+			return byteLen, nil
+		}
+		elemByteLen, err := t.childTypes[0].ByteLen()
+		if err != nil {
+			return -1, err
+		}
+		return int(t.staticLength) * elemByteLen, nil
+	case Tuple:
+		size := 0
+		for i := 0; i < len(t.childTypes); i++ {
+			if t.childTypes[i].enumIndex == Bool {
+				// search previous bool
+				before := findBoolLR(t.childTypes, i, -1)
+				// search after bool
+				after := findBoolLR(t.childTypes, i, 1)
+				// append to heads and tails
+				if before%8 != 0 {
+					return -1, fmt.Errorf("expected before has number of bool mod 8 = 0")
+				}
+				if after > 7 {
+					after = 7
+				}
+				i += after
+				size++
+			} else {
+				childByteSize, err := t.childTypes[i].ByteLen()
+				if err != nil {
+					return -1, err
+				}
+				size += childByteSize
+			}
+		}
+		return size, nil
+	default:
+		return -1, fmt.Errorf("%s is a dynamic type", t.String())
+	}
+}
+
+// Value struct is the ABI Value, holding ABI Type information and the ABI value representation.
+type Value struct {
+	ABIType Type
+	value   interface{}
+}
```

Comment:
I make ABI type public, but I am not sure if `value` field should be public, probably need to discuss somewhat about that.